File: ios/UmbraCore/Sources/UmbraCore/WireGuard/WgConfigStore.swift
import Foundation

/// Stores the imported WireGuard config on disk — the iOS analogue of the Android
/// WgConfigStore and the phone's `/etc/wireguard/vpn.conf`. The directory is
/// injected (the app passes its App Group container so the Packet Tunnel
/// extension can read it too), which also makes this unit-testable.
public struct WgConfigStore {
    public static let name = "vpn"

    private let fileURL: URL

    public init(directory: URL) {
        self.fileURL = directory.appendingPathComponent("\(Self.name).conf")
    }

    /// Validate then persist. Throws WgConfigError on a bad config.
    public func save(_ text: String) throws {
        _ = try WgConfig.validate(text)
        try FileManager.default.createDirectory(
            at: fileURL.deletingLastPathComponent(), withIntermediateDirectories: true)
        try text.write(to: fileURL, atomically: true, encoding: .utf8)
    }

    public func load() -> String? { try? String(contentsOf: fileURL, encoding: .utf8) }

    public func exists() -> Bool { FileManager.default.fileExists(atPath: fileURL.path) }

    public func summary() -> WgConfigSummary? {
        guard let text = load() else { return nil }
        return try? WgConfig.validate(text)
    }

    public func clear() { try? FileManager.default.removeItem(at: fileURL) }
}

File: android/app/src/main/java/com/forrestlasiter/umbra/wg/WgConfigStore.kt
package com.forrestlasiter.umbra.wg

import android.content.Context
import java.io.File

/**
 * Stores the imported WireGuard config in app-private storage — the phone's
 * `/etc/wireguard/vpn.conf`. Only this app can read it (Android's per-app
 * sandbox); it never leaves the device.
 */
class WgConfigStore(context: Context) {

    private val dir = File(context.filesDir, "wireguard").apply { mkdirs() }
    private val file = File(dir, "$NAME.conf")

    fun save(text: String) {
        WgConfig.validate(text)              // never store a config we couldn't parse
        file.writeText(text)
    }

    fun load(): String? = if (file.exists()) file.readText() else null

    fun summary(): WgConfigSummary? = load()?.let { runCatching { WgConfig.validate(it) }.getOrNull() }

    fun exists(): Boolean = file.exists()

    fun clear() { file.delete() }

    companion object {
        const val NAME = "vpn"               // matches the Linux `umbra vpn --name vpn`
    }
}
